File: apps/web/src/lib/utils/gemini.ts
// lib/services/gemini.ts
import { GoogleGenerativeAI } from "@google/generative-ai";

export class GeminiService {
  private genAI: GoogleGenerativeAI;

  constructor(apiKey: string) {
    this.genAI = new GoogleGenerativeAI(apiKey);
  }

  // async generateResponse(message: string): Promise<string> {
  //   const model = this.genAI.getGenerativeModel({
  //     model: "gemini-2.0-flash",
  //     generationConfig: {
  //       temperature: 1,
  //       topP: 0.95,
  //       topK: 40,
  //       maxOutputTokens: 8192,
  //       responseMimeType: "text/plain",
  //     },
  //   });

  //   const result = await model.generateContent(message);
  //   const response = await result.response;
  //   return response.text();
  // }

  async streamResponse(
    chatId: string,
    message: string,
    onToken: (token: string) => Promise<void>,
  ): Promise<void> {
    try {
      const model = this.genAI.getGenerativeModel({
        model: "gemini-2.0-flash",
        generationConfig: {
          temperature: 1,
          topP: 0.95,
          topK: 40,
          maxOutputTokens: 8192,
          responseMimeType: "text/plain",
        },
      });

      const result = await model.generateContentStream(message);
      let buffer = "";

      for await (const chunk of result.stream) {
        const text = chunk.text();
        if (text) {
          // Split the text into words and process each
          const words = text.split(/(\s+)/);
          for (const word of words) {
            buffer += word;
            // Send buffer when we have a complete word or punctuation
            if (word.match(/[\s.!?,]/)) {
              await onToken(buffer);
              buffer = "";
            }
          }
        }
      }

      // Send any remaining buffer
      if (buffer) {
        await onToken(buffer);
      }
    } catch (error) {
      console.error("Gemini streaming error:", error);
      throw error;
    }
  }
}

const geminiService = new GeminiService(
  process.env.NEXT_PUBLIC_GEMINI_API_KEY!,
);
export default geminiService;

File: services/compute/zephyr_rust_service/src/ai/gemini.rs
// src/ai/gemini.rs
use super::AIProvider;
use anyhow::Result;
use async_trait::async_trait;
use serde_json::json;
use dotenv::dotenv;
use serde_json::Value;
use std::sync::Arc;
use tokio::sync::Mutex;

pub struct GeminiProvider {
    api_key: String,
    base_url: String,
}

#[async_trait]
impl AIProvider for GeminiProvider {
    async fn stream_response<F>(&self, content: &str, mut callback: F) -> Result<()>
    where
        F: FnMut(String) + Send + 'static,
    {
        let client = reqwest::Client::new();
        let url = format!("{}{}", self.base_url, self.api_key);

        let request_body = json!({
            "contents": [{
                "parts": [{
                    "text": content
                }]
            }],
            "generationConfig": {
                "temperature": 0.7,
                "topK": 40,
                "topP": 0.95,
                "maxOutputTokens": 2048,
            }
        });

        println!("🔄 Making request to Gemini");

        let response = client
            .post(&url)
            .json(&request_body)
            .send()
            .await?;

        let status = response.status();
        if !status.is_success() {
            let error_text = response.text().await?;
            println!("❌ Gemini API error: {} - {}", status, error_text);
            return Err(anyhow::anyhow!("API request failed: {}", error_text));
        }

        let response_text = response.text().await?;
        println!("📝 Full Gemini response: {}", response_text);

        if let Ok(json) = serde_json::from_str::<Value>(&response_text) {
            if let Some(candidates) = json["candidates"].as_array() {
                if let Some(first_candidate) = candidates.first() {
                    if let Some(text) = first_candidate["content"]["parts"][0]["text"].as_str() {
                        println!("✨ Processing Gemini response: {}", text);

                        // Create a buffer for accumulating words
                        let mut word_buffer = String::new();

                        for c in text.chars() {
                            word_buffer.push(c);

                            // If we hit a word boundary (space or punctuation), send the token
                            if c.is_whitespace() || c.is_ascii_punctuation() {
                                if !word_buffer.is_empty() {
                                    println!("🔤 Sending token: {}", word_buffer);
                                    callback(word_buffer.clone());
                                    word_buffer.clear();
                                }
                            }
                        }

                        // Send any remaining content in the buffer
                        if !word_buffer.is_empty() {
                            println!("🔤 Sending final token: {}", word_buffer);
                            callback(word_buffer);
                        }

                        println!("✅ Finished processing Gemini response");
                    } else {
                        println!("❌ No text found in response parts");
                        return Err(anyhow::anyhow!("No text found in response parts"));
                    }
                } else {
                    println!("❌ No candidates found in response");
                    return Err(anyhow::anyhow!("No candidates found in response"));
                }
            }
        }

        Ok(())
    }

    async fn process_message(&self, content: &str) -> Result<String> {
        let client = reqwest::Client::new();
        let url = format!("{}{}", self.base_url, self.api_key);

        let request_body = json!({
            "contents": [{
                "parts": [{
                    "text": content
                }]
            }],
            "generationConfig": {
                "temperature": 0.7,
                "topK": 40,
                "topP": 0.95,
                "maxOutputTokens": 2048,
            }
        });

        let response = client
            .post(&url)
            .json(&request_body)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await?;
            return Err(anyhow::anyhow!("API request failed: {}", error_text));
        }

        let response_text = response.text().await?;
        Ok(response_text)
    }
}

impl GeminiProvider {
    pub fn new() -> Result<Self> {
        dotenv().ok();
        let api_key = std::env::var("GEMINI_API_KEY")
            .expect("GEMINI_API_KEY must be set");

        Ok(Self {
            api_key,
            base_url: "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=".to_string(),
        })
    }
}
